kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/bootstrap/FilterChainFactory.java

```diff
@@ -78,20 +99,47 @@ public ScheduledExecutorService eventLoop() {
                     return pfr.pluginFactory(pluginClass).pluginInstance(instanceName);
                 }
             };
-            this.initialized = filterDefinitions.stream()
-                    .map(fd -> {
-                        FilterFactory<? super Object, ? super Object> filterFactory = pluginFactory.pluginInstance(fd.type());
-                        Class<?> configType = pluginFactory.configType(fd.type());
-                        if (fd.config() == null || configType.isInstance(fd.config())) {
-                            return new UninitializedFilterFactory(fd.type(), filterFactory, fd.config());
-                        }
-                        else {
-                            throw new PluginConfigurationException("accepts config of type " +
-                                    configType.getName() + " but provided with config of type " + fd.config().getClass().getName() + "]");
-                        }
-                    })
-                    .map(uff -> uff.initialize(context))
-                    .toList();
+            this.initialized = new ArrayList<>(filterDefinitions.size());
+            try {
+                for (var fd : filterDefinitions) {
+                    FilterFactory<? super Object, ? super Object> filterFactory = pluginFactory.pluginInstance(fd.type());
+                    Class<?> configType = pluginFactory.configType(fd.type());
+                    if (fd.config() == null || configType.isInstance(fd.config())) {
+                        Wrapper uninitializedFilterFactory = new Wrapper(context, fd.type(), filterFactory, fd.config());
+                        this.initialized.add(uninitializedFilterFactory);
+                    }
+                    else {
+                        throw new PluginConfigurationException("accepts config of type " +
+                                configType.getName() + " but provided with config of type " + fd.config().getClass().getName() + "]");
+                    }
+                }
+            }
+            catch (Exception e) {
+                // close already initialized factories
+                close();
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        RuntimeException firstThrown = null;
+        for (Wrapper wrapper : initialized) {
```

Comment:
I think we should close the factories in the reverse order to which they were initialised.     We should document that behaviour too.

Reply:
I don't think it should be a documented part of the `FilterFactory` contract (if that's where you were suggesting to document it), because we don't really want to get into there being dependencies/shared state between filter factories, do we? I certainly can't think of a use case for it.